### PR TITLE
Bump protobuf to v24.4 and update generator script

### LINKED ISSUE #11685: Type-stub for CodeGeneratorRequest in google.protobuf
## Issue Summary

I thought the `google.protobuf.compiler.plugin_pb2.CodeGeneratorRequest` type-stub did not match the actual code and would like to resolve this.

```bash
pip install mypy==1.9.0 protobuf==4.24.0 types-protobuf==4.24.0.20240311
```

```python
# testcode.py
from google.protobuf.compiler.plugin_pb2 import CodeGeneratorRequest

request = CodeGeneratorRequest()
assert "source_file_descriptors" in {field.name for field in request.DESCRIPTOR.fields}
print(f"{request.source_file_descriptors=}")
```

```bash
$ python testcode.py  # Correct runtime code
request.source_file_descriptors=[]
$ mypy testcode.py  # Incorrect type checking due to incorrect type hints
testcode.py:6: error: "CodeGeneratorRequest" has no attribute "source_file_descriptors"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

### Proposal

Put `source_file_descriptors` property to type-stub for `google.protobuf.compiler.plugin_pb2.CodeGeneratorRequest`

I have never done PR on this project so any advice would be appreciated.

### PR DESCRIPTION
**Pull request by Avasam**
Closes #11686 and closes #11685

As of v22, protobuf downloads on github no longer provides a zip per language, but rather a single zip with all languages. I've updated the script to work on v22+ and bumped protobuf to v24 to reflect what the stubs claim to cover.

I could easily have the `version` in `METADATA.toml` follow the `python protobuf` version used by the script. LMK if you also think that would be a good idea (with an appropriate comment in `METADATA.toml`)

I didn't bump all the way up to v26 (#10964) because I wanted to at least correct the existing v24 typings

CC @AlexWaygood  I also fixed the `pre-commit run --files` invocation so it's actually passed the files to check, runs without error, and re-run Ruff as a second pass because unused imports are not removed when fixing `UP036 Version block is outdated for minimum Python version`

CC @ijknabla FYI

### CODE REVIEW

File: scripts/sync_tensorflow_protobuf_stubs.sh [hunk 1]
```diff
@@ -61,13 +61,19 @@ rm tensorflow/compiler/xla/service/hlo_execution_profile_data_pb2.pyi \
...
 rm -rf repository/
+
+# Must be run in a git repository
+cd $REPO_ROOT
+# use `|| true` so the script still continues even if a pre-commit hook
+# applies autofixes (which will result in a nonzero exit code)
+pre-commit run --files $(git ls-files -- "$REPO_ROOT/stubs/tensorflow/tensorflow") || true
+# Ruff takes two passes to fix everything, re-running all of pre-commit is *slow*
+# and we don't need --unsafe-fixes to remove imports
+ruff check "$REPO_ROOT/stubs/protobuf" --fix --exit-zero
```
Comment (Avasam) [inline scripts/sync_tensorflow_protobuf_stubs.sh:79]:
```suggestion
ruff check "$REPO_ROOT/stubs/tensorflow/tensorflow" --fix --exit-zero
```

File: scripts/generate_proto_stubs.sh [hunk 2]
```diff
@@ -70,10 +69,17 @@ PROTO_FILES=$(grep "GenProto.*google" $PYTHON_PROTOBUF_DIR/python/setup.py | \
...
 protoc_install/bin/protoc --proto_path="$PYTHON_PROTOBUF_DIR/src" --mypy_out="relax_strict_optional_primitives:$REPO_ROOT/stubs/protobuf" $PROTO_FILES
 
-# use `|| true` so the script still continues even if a pre-commit hook
-# applies autofixes (which will result in a nonzero exit code)
-pre-commit run --files "$REPO_ROOT/stubs/protobuf" || true
+PYTHON_PROTOBUF_VERSION=$(jq -r '.[] | .languages.python' "$PYTHON_PROTOBUF_DIR/version.json")
 
 sed --in-place="" \
-  "s/extra_description = .*$/extra_description = \"Generated using [mypy-protobuf==$MYPY_PROTOBUF_VERSION](https:\/\/github.com\/nipunn1313\/mypy-protobuf\/tree\/v$MYPY_PROTOBUF_VERSION) on protobuf==$PYTHON_PROTOBUF_VERSION\"/" \
+  "s/extra_description = .*$/extra_description = \"Generated using [mypy-protobuf==$MYPY_PROTOBUF_VERSION](https:\/\/github.com\/nipunn1313\/mypy-protobuf\/tree\/v$MYPY_PROTOBUF_VERSION) on [protobuf v$PROTOBUF_VERSION](https:\/\/github.com\/protocolbuffers\/protobuf\/releases\/tag\/v$PROTOBUF_VERSION) (python protobuf==$PYTHON_PROTOBUF_VERSION)\"/" \
```
Comment (Avasam) [inline scripts/generate_proto_stubs.sh:75]: This is a very long line, but idk how to do a single argument multiline string in bash

File: stubs/protobuf/google/protobuf/api_pb2.pyi [hunk 3]
```diff
@@ -214,7 +214,7 @@ class Mixin(google.protobuf.message.Message):
     The mixin construct implies that all methods in `AccessControl` are
     also declared with same name and request/response types in
     `Storage`. A documentation generator or annotation processor will
-    see the effective `Storage.GetAcl` method after inheriting
+    see the effective `Storage.GetAcl` method after inherting
```
Comment (AlexWaygood) [inline stubs/protobuf/google/protobuf/api_pb2.pyi:217]: Lol, did they deliberately introduce a typo or did we accidentally update this in our vendored copy without contributing the typo fix upstream?

**Comment by AlexWaygood** [comment]
> CC @AlexWaygood I also fixed the `pre-commit run --files` invocation so it's actually passed the files to check, runs without error, and re-run Ruff as a second pass because unused imports are not removed when fixing `UP036 Version block is outdated for minimum Python version`

Thanks, those changes look great :)

File: stubs/protobuf/google/protobuf/api_pb2.pyi (hunk 3 above)
Comment (Avasam) [reply stubs/protobuf/google/protobuf/api_pb2.pyi:217]: They deliberately re-introduced it! https://github.com/protocolbuffers/protobuf/commit/b81c127e94d953aa9be5c9f00a1ad323552bae0e

Comment (AlexWaygood) [reply stubs/protobuf/google/protobuf/api_pb2.pyi:217]: Amazing! If this didn't have a citation I'd be certain this was an April Fool's :D

Comment (JelleZijlstra) [reply stubs/protobuf/google/protobuf/api_pb2.pyi:217]: Got to have copyright on those typo fixes

File: scripts/generate_proto_stubs.sh (hunk 2 above)
Comment (ijknabla) [reply scripts/generate_proto_stubs.sh:75]: I think you can split a line by placing `”¥¥¥n”`  immediately after the word you want to separate.

```
"s/extra_description = .*$/extra_description = "\
"\"Generated using [mypy-protobuf==$MYPY_PROTOBUF_VERSION](https:\/\/github.com\/nipunn1313\/mypy-protobuf\/tree\/v$MYPY_PROTOBUF_VERSION)"\
" on [protobuf v$PROTOBUF_VERSION](https:\/\/github.com\/protocolbuffers\/protobuf\/releases\/tag\/v$PROTOBUF_VERSION)"\
" (python protobuf==$PYTHON_PROTOBUF_VERSION)\"/"
```

File: stubs/protobuf/google/protobuf/compiler/plugin_pb2.pyi [hunk 4]
```diff
@@ -101,6 +104,12 @@ class CodeGeneratorRequest(google.protobuf.message.Message):
         fully qualified.
         """
     @property
+    def source_file_descriptors(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[google.protobuf.descriptor_pb2.FileDescriptorProto]:
```
Comment (ijknabla) [inline stubs/protobuf/google/protobuf/compiler/plugin_pb2.pyi:107]: This code could solve the problem I pointed out in https://github.com/python/typeshed/issues/11685 . (We can close https://github.com/python/typeshed/issues/11685 .)

**Pull request merged**